### PR TITLE
Improve EVM test

### PR DESCRIPTION
**Pull request by leonardoalt**
This PR:

- fixes the EVM test by executing the tx after setting it up
- changes the EVM test to receive the program's bytecode via prover input instead of hardcoded
- removes `degree` statements from std machines
- changes the riscv compiler to auto detect the degree needed by the program

### CODE REVIEW

File: riscv/tests/riscv_data/evm/src/lib.rs [hunk 1]
```diff
@@ -24,17 +19,18 @@ fn main() {
         b256!("e3c84e69bac71c159b2ff0d62b9a5c231887a809a96cb4a262a4b96ed78a1db2");
     let mut db = CacheDB::new(EmptyDB::default());
 
+    let bytecode_len = get_prover_input(0);
+    let bytecode: Vec<_> = (1..(bytecode_len + 1)).map(|idx| get_prover_input(idx) as u8).collect();
+
     // Fill database:
-    let bytecode = Bytes::from_static(BYTECODE);
+    let bytecode = Bytes::from(bytecode);
```
Comment (georgwiese) [inline riscv/tests/riscv_data/evm/src/lib.rs:26]: Does that work with `Vec<usize>` as well? Then you need to call `get_prover_input()` 4x fewer times. But if bytecodes are short, this is probably fine.

File: riscv/src/compiler.rs [hunk 2]
```diff
@@ -185,29 +185,45 @@ pub fn compile(mut assemblies: BTreeMap<String, String>, coprocessors: &CoProces
...
+                .into_iter()
+                .flat_map(|v| process_statement(v, coprocessors)),
+        )
+        .chain(["// This is the data initialization routine.\n__data_init::".to_string()])
+        .chain(data_code)
+        .chain(["// This is the end of the data initialization routine.\nret;".to_string()])
+        .collect();
+
+    // The program ROM needs to fit the degree, so we use the next power of 2.
+    let degree = u64::BITS - (program.len() as u64).leading_zeros() + 1;
```
Comment (georgwiese) [inline riscv/src/compiler.rs:209]:
```suggestion
    let degree = program.len().ilog2() + 1;
```

File: riscv/src/compiler.rs [hunk 3]
```diff
@@ -185,29 +185,45 @@ pub fn compile(mut assemblies: BTreeMap<String, String>, coprocessors: &CoProces
...
+    // In practice, these are the lengths of single proofs that we want to support.
+    // Reasoning:
+    // - 18: is the lower bound for the Binary and Shift machines.
+    // - 20: revm's ROM does not fit in 2^19.
+    // - >20: may be needed in the future.
+    // This is an assert for now, but could be a compiler warning or error.
+    // TODO note that if the degree is higher than 18 we might need mux machines for Binary and
+    // Shift.
+    assert!((18..=20).contains(&degree));
+    let degree = 1 << degree;
```
Comment (georgwiese) [inline riscv/src/compiler.rs:220]: Maybe log it?

File: riscv/tests/riscv_data/evm/src/lib.rs (hunk 1 above)
Comment (leonardoalt) [reply riscv/tests/riscv_data/evm/src/lib.rs:26]: Yea for now they're short so I just wanted to keep it simple. But yea makes sense to do that eventually

File: riscv/src/compiler.rs (hunk 3 above)
Comment (leonardoalt) [reply riscv/src/compiler.rs:220]: what exactly? what the degree is?

Comment (georgwiese) [reply riscv/src/compiler.rs:220]: Sorry, yes! Think it's pretty important information as it directly influences the proving time and hard to estimate by the user.

Comment (leonardoalt) [reply riscv/src/compiler.rs:222]: done

File: riscv/src/compiler.rs (hunk 2 above)
Comment (leonardoalt) [reply riscv/src/compiler.rs]: done

File: riscv/src/compiler.rs [hunk 4]
```diff
@@ -185,29 +185,47 @@ pub fn compile(mut assemblies: BTreeMap<String, String>, coprocessors: &CoProces
...
+        )
+        .chain(["// This is the data initialization routine.\n__data_init::".to_string()])
+        .chain(data_code)
+        .chain(["// This is the end of the data initialization routine.\nret;".to_string()])
+        .collect();
+
+    // The program ROM needs to fit the degree, so we use the next power of 2.
+    let degree = program.len().ilog2() + 1;
+    let degree = std::cmp::max(degree, 18);
+    log::info!("Inferred degree {degree}");
```
Comment (georgwiese) [inline riscv/src/compiler.rs:211]:
```suggestion
    log::info!("Inferred degree 2^{degree}");
```

**Review by georgwiese — APPROVED** [review]

**Review by chriseth — APPROVED** [review]

**Review by georgwiese — APPROVED** [review]

**Comment by leonardoalt** [comment]
Actually had to comment out the test again otherwise the tests are gonna fail, because the compiler and witgen need too much RAM to run this test.

**Pull request merged**